packages/satcheljs/test/createUndoTests.ts

```diff
@@ -0,0 +1,370 @@
+
+            let object = observable({key: oldValue});
+            let undoableAction = action('updateObject')(() => { object.key = newValue });
+
+            let undoResult = createUndo('updateObject', true)(undoableAction);
+            action('updateObject-again')(() => { object.key = 100 })();
+
+            expect(undoResult).toThrow();
+        });
+
+        it('throws an exception when it undoes an array splice', () => {
+            let object = observable([1, 2, 3, 4, 5, 6] as any[]);
+            let undoableAction = action('spliceArray')(() => { object.splice(2, 3, 'a') });
+
+            let undoResult = createUndo('spliceArray', true)(undoableAction);
+            action('spliceArray-again')(() => { object[2] = 100 })();
+
+            expect(undoResult).toThrow();
+        });
+
+        it('throws an exception when it undoes an add to a map', () => {
+            let index = 'key';
+            let newValue = 5;
+
+            let object = map({});
+            let undoableAction = action('addMap')(() => { object.set(index, newValue) });
+
+            let undoResult = createUndo('addMap', true)(undoableAction);
+            action('addMap-again')(() => {object.set(index, 100);})();
+
+            expect(undoResult).toThrow();
+        });
+
+        it('throws an exception when it undoes an add to an object', () => {
+            let index = 'key';
+            let newValue = 5;
+
+            let object: any = observable({});
+            let undoableAction = action('addObject')(() => { extendObservable(object, {[index]: newValue}) });
+
+            let undoResult = createUndo('addObject', true)(undoableAction);
+            action('addMap-again')(() => {object[index] = 100;})();
+
+            expect(undoResult).toThrow();
+        });
+
+        it('throws an exception when it undoes a delete to a map', () => {
+            let index = 'key';
+            let oldValue = 5;
+
+            let object = map({[index]: oldValue});
+            let undoableAction = action('deleteMap')(() => { object.delete(index) });
+
+            let undoResult = createUndo('deleteMap', true)(undoableAction);
+            action('deleteMap-again')(() => {object.set(index, 100);})();
+
+            expect(undoResult).toThrow();
+        });
+    });
+
+    it('handles nested undo windows', () => {
+        let array = observable([1, 2, 3, 4, 5]);
+
+        let outerUndo = createUndo('outerUndo')(action('outerUndo')(() => {
+            array[0] = 0;
+
+            expect(array.slice(0)).toEqual([0, 2, 3, 4, 5]);
+
+            let innerUndo = createUndo('innerUndo')(action('innerUndo')(() => {
+                array[1] = 0;
+            }));
+
+            array[2] = 0;
+
+            expect(array.slice(0)).toEqual([0, 0, 0, 4, 5]);
+
+            innerUndo();
+
+            expect(array.slice(0)).toEqual([0, 2, 0, 4, 5]);
+        }));
+
+        outerUndo();
+
+        expect(array.slice(0)).toEqual([1, 2, 3, 4, 5]);
+    });
+});
```

Comment:
I've made a change to throw if an instance is called multiple times, and have added a test case for it.